### ISSUE TITLE
Multiple background threads spawned

### ISSUE THREAD

**Issue by ktoh** [miguelgrinberg/Flask-SocketIO#126]
I'm using the flask-socketio with a background thread similar to the [example app](https://github.com/miguelgrinberg/Flask-SocketIO/blob/master/example/app.py#L17).

The thread is configured as follows:

```
# Initialize app
from flask_socketio import SocketIO
socketio = SocketIO(app)

# Set up background thread
from gevent import monkey, sleep
monkey.patch_all()

import threading

def background_thread():
    while True:
        task = task_queue.pop()   # pop a task off the queue
        if task:
            do_task(task, socketio)
            logger.debug('Task {} is handled by thread {}'.format(task, threading.current_thread())
        sleep(2)

# Initialize the thread once per app
if getattr(app, 'background_thread', None) is None:
    thread = threading.Thread(target=background_thread)
    thread.start()
    app.background_thread = thread

```

I'm finding that after `thread.start()` is called, there are TWO threads, a <Thread> instance, and a <_DummyThread> instance.
As a result, the task might get handled by either the _DummyThread or the Thread instance.

Moreover:

1. **the `socketio` instance is different depending on the thread handling the task**.  So the task results are unpredictable.
2. The `_DummyThread` instance seems to contain the correct `socketio` instance (i.e. its namespaces and rooms update as requests come in and out).  The `Thread` instance doesn't seem to update its `socketio` object, and so tasks fail when they happen to get served by this thread.

I'm not sure why there are two threads of differnet classes spawned upon `thread.start()`.  Any ideas why this might be the case?

**Comment by miguelgrinberg** [comment]
Try moving the monkey patching to the very top of your file, above the import statements.

**Comment by ktoh** [comment]
Thanks for the quick response!

I did try hoisting monkey.patch_all() to the very top so that it's the first line of code gunicorn sees.  It's still spawning the two threads.

I'm using:
* Gunicorn 18.0
* gevent 1.0.1
* python 2.7.3
* socketio 0.3.5

**Comment by miguelgrinberg** [comment]
Sorry, I sort of dropped the ball on this. Did you figure it out?

I think the fact that you are seeing two socketio instances means that you have two processes, not two threads. If you are still having trouble with this, please show me the output of gunicorn.

**Comment by ktoh** [comment]
@miguelgrinberg  I didn't resolve it entirely, but given the significant improvement available with flask-socketio v1.0, I will probably migrate to the eventlet-based architecture and drop gevent (and the brittle dependency on gevent-socketio) altogether.

That seems like a better approach to addressing this issue, which I suspect is the result of some crappy version bug with antiquited gunicorn, gevent or socketio versions.

I did patch my problem for now by introspecting the thread and executing tasks only in the correct thread, but that's not a real solution.  Will report back after moving to eventlet in the next few weeks.

thanks much

**Comment by ktoh** [comment]
Upgraded to flask-socketio 1.0b running with eventlets and this problem went away as expected. 

I'm closing the issue accordingly. I never quite narrowed down the problem, but given the reliance of gevent-websocket on stale socketio and gevent versions it just isn't worth tracking down given that 1.0 provides better alternatives to gevent.
